Amazon S3 Media Storage support

neglectedvalue opened the pull request:
Hi, long time ago I've created module to support AWS S3 bucket for Orchard media storage and since OrchardCore was in my opinion Azure based I did not think about any contributions, but recently @Piedone contacted with me and suggested to create a PR.
Basically nothing complex, module for Azure Media support was done nicely and understandable so I've made mine as much equal as possible.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=11738) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=11738) before we can accept your contribution.<br/><br/>:x: neglectedvalue [sign now](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=11738)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/OrchardCMS/OrchardCore?pullRequest=11738) it.</sub>

hishamco commented:
I was planned to make this for Orchard Core Contrib (OCC), but it's good to see someone push a PR to implement it

jtkech commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/Manifest.cs:
```diff
@@ -0,0 +1,19 @@
+)]
+
+[assembly: Feature(
+    Id = "OrchardCore.Media.AmazonS3.Storage",
```
@neglectedvalue Just for info

When a module doesn't define any feature, in fact it is like it defines only one feature whose Id equals to the module Id (the assembly name most of the time equals to the project/package name), here `OrchardCore.Media.AmazonS3`.

When a module define at least one feature, we are used to set the Id of the first feature to the module assembly name, one difference is that for some definitions as Controlers, Migrations, Startups (as in your case) you then don't need to decorate them with the `[Feature("...")]` attribute.

But yes, the existing `OrchardCore.Media.Azure` doesn't follow this "rule", maybe this is the only one ;)

So as you want, this was just for info.

hishamco replied:
> But yes, the existing OrchardCore.Media.Azure doesn't follow this "rule", maybe this is the only one ;)

I may check the other once and fix

jtkech replied:
Thanks for the update

For the module attribute you don't need to set the Id, it is modifiable for people defining a different assembly name in their project file, so here would need to be `Id = "OrchardCore.Media.AmazonS3"` but useless.

But always better to define a feature, even if there is only one, so that it is simpler if you add more features.

**So you can just keep what you did before but just replace for this only one feature** (so the 1st one) this

    Id = "OrchardCore.Media.AmazonS3.Storage",

with

    Id = "OrchardCore.Media.AmazonS3",

jtkech replied:
using OrchardCore.Modules.Manifest;

    [assembly: Module(
        Name = "Amazon S3 Media",
        Author = ManifestConstants.OrchardCoreTeam,
        Website = ManifestConstants.OrchardCoreWebsite,
        Version = ManifestConstants.OrchardCoreVersion
    )]

    [assembly: Feature(
        Id = "OrchardCore.Media.AmazonS3",
        Name = "Amazon S3 Media Storage",
        Description = "Enables support for storing media files in Amazon S3 Bucket.",
        Dependencies = new[]
        {
            "OrchardCore.Media.Cache"
        },
        Category = "Hosting"
    )]

neglectedvalue replied:
Yeap, done, thx for that info.

Piedone commented on src/OrchardCore.Cms.Web/appsettings.json:
```diff
@@ -129,6 +129,20 @@
+    //"OrchardCore_Media_Amazon_S3": {
+    //  "BucketName": "",
+    // Credentials section needed only if Orchard will be hosted not in the AWS Cloud
+    // since in the AWS you dont need to store this info in the appsettings file, you 
```
```suggestion
    // since in the AWS you don't need to store this info in the appsettings file, you 
```

Piedone commented on src/OrchardCore.Cms.Web/appsettings.json:
```diff
@@ -129,6 +129,20 @@
-    //}
+    //},
+    
+    //"OrchardCore_Media_Amazon_S3": {
+    //  "BucketName": "",
+    // Credentials section needed only if Orchard will be hosted not in the AWS Cloud
+    // since in the AWS you dont need to store this info in the appsettings file, you 
+    // just need to set BucketName and BasePath.
+    //    "Credentials": {
+    //      "SecretKey": "",
+    //      "AccessKeyId": "",
+    //      "RegionEndpoint": ""
+    //    },
+    //    "BasePath": "/media",
+    //    "ProfileName": ""
+    //  }
```
Move this up so the settings are in alphabetical order.

Piedone commented on src/OrchardCore.Cms.Web/appsettings.json:
```diff
@@ -129,6 +129,20 @@
-    //}
+    //},
+    
+    //"OrchardCore_Media_Amazon_S3": {
```
Add a docs page for this like https://docs.orchardcore.net/en/latest/docs/reference/modules/Media.Azure/, [see contribution notes](https://docs.orchardcore.net/en/latest/docs/topics/docs-contributions/). Link it here like the others are linked.

Piedone commented on src/OrchardCore/OrchardCore.FileStorage.AmazonS3/AmazonWebServiceResponseExtensions.cs:
```diff
@@ -0,0 +1,13 @@
+    public static bool IsSuccessful(this AmazonWebServiceResponse response) =>
+        response.HttpStatusCode == HttpStatusCode.OK;
+
+    public static bool DeletedSuccessfully(this AmazonWebServiceResponse response) =>
```
Just to be in line with `IsSuccessful`:

```suggestion
    public static bool IsDeleteSuccessful(this AmazonWebServiceResponse response) =>
```

Piedone commented on src/OrchardCore/OrchardCore.FileStorage.AmazonS3/AwsDirectory.cs:
```diff
@@ -0,0 +1,32 @@
+        _path = path;
+        _lastModifiedUtc = lastModifiedUtc;
+        _name = System.IO.Path.GetFileName(path);
+        _directoryPath = _path.Length > _name.Length ?
+            _path.Substring(0, _path.Length - _name.Length - 1) : String.Empty;
```
```suggestion
        _directoryPath = _path.Length > _name.Length ? _path[..(_path.Length - _name.Length - 1)] : String.Empty;
```

Piedone commented on src/OrchardCore/OrchardCore.FileStorage.AmazonS3/AwsFile.cs:
```diff
@@ -0,0 +1,42 @@
+        }
+        else
+        {
+            _directoryPath = _path.Substring(0, _path.Length - _name.Length - 1);
```
```suggestion
            _directoryPath = _path[..(_path.Length - _name.Length - 1)];
```

Piedone commented on src/OrchardCore/OrchardCore.FileStorage.AmazonS3/AwsFile.cs:
```diff
@@ -0,0 +1,42 @@
+        _path = path;
+        _name = System.IO.Path.GetFileName(_path);
+
+        if (_name == _path) // file is in root Directory
+        {
+            _directoryPath = String.Empty;
+        }
```
```suggestion
        if (_name == _path)
        {
            // File is in the root directory.
            _directoryPath = String.Empty;
        }
```

Piedone commented on src/OrchardCore/OrchardCore.FileStorage.AmazonS3/AwsStorageOptions.cs:
```diff
@@ -0,0 +1,49 @@
+    public string BucketName { get; set; }
+
+    /// <summary>
+    /// The base directory path to use inside the container for this stores contents.
```
```suggestion
    /// The base directory path to use inside the container for this store's contents.
```

Piedone commented on src/OrchardCore/OrchardCore.FileStorage.AmazonS3/AwsStorageOptions.cs:
```diff
@@ -0,0 +1,49 @@
+    /// <summary>
+    /// Gets or sets the credentials.
+    /// <remarks>
+    /// Credentials section can be set directly via configuration or get loaded from the configured ProfileName.
+    /// For development purpose it is recommended to specify just ProfileName.
+    /// For Prod Env this section should be null, AWS SDK Services will get the default credentials from Env variables
```
```suggestion
    /// Credentials section can be set directly via configuration or get loaded from the configured ProfileName.
    /// For development purposes, it is recommended to specify just ProfileName.
    /// For a production environment this section should be null, AWS SDK Services will get the default credentials
    /// from environment variables.
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsExtension.cs:
```diff
@@ -0,0 +1,90 @@
+        }
+        else
+        {
+            // Attempt to load Credentials from Profile
```
```suggestion
            // Attempt to load Credentials from Profile.
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/Startup.cs:
```diff
@@ -0,0 +1,170 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+using Amazon;
+using Amazon.Runtime;
+using Amazon.S3;
+
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Configuration;
+using OrchardCore.FileStorage;
+using OrchardCore.Media.Core;
+using OrchardCore.Media.Events;
+using OrchardCore.Modules;
+using OrchardCore.FileStorage.AmazonS3;
+using OrchardCore.Media.Core.Events;
```
```suggestion
using System;
using System.IO;
using System.Linq;
using System.Text;
using Amazon;
using Amazon.Runtime;
using Amazon.S3;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using OrchardCore.Environment.Shell;
using OrchardCore.Environment.Shell.Configuration;
using OrchardCore.FileStorage;
using OrchardCore.FileStorage.AmazonS3;
using OrchardCore.Media.Core;
using OrchardCore.Media.Core.Events;
using OrchardCore.Media.Events;
using OrchardCore.Modules;
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/Startup.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+        else
+        {
+            _logger.LogInformation(
+                $"Starting with S3 Media Configuration. {System.Environment.NewLine} BucketName: {storeOptions.BucketName}, {System.Environment.NewLine} BasePath: {storeOptions.BasePath}");
```
Parameterize the logging template.

Piedone replied:
This is still not addressed :). What I'm telling is, don't add variables in the string, but supply them as parameters, like you do on line 51.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/Startup.cs:
```diff
@@ -0,0 +1,170 @@
+
+                if (String.IsNullOrWhiteSpace(hostingEnvironment.WebRootPath))
+                {
+                    throw new Exception("The wwwroot folder for serving cache media files is missing.");
```
Perhaps there is a less generic exception appropriate to throw here?

neglectedvalue replied:
Yes, I've also thought about it, should I add smth like MediaConfigurationException into the Media.Abstractions?

Piedone replied:
Sounds good, yes.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/Startup.cs:
```diff
@@ -0,0 +1,170 @@
+                stringBuilder.Append(error.ErrorMessage);
+            }
+
+            if (_webHostEnvironment.IsDevelopment())
+            {
+                _logger.LogInformation(
+                    $"S3 Media configuration validation failed: {validationErrors} fallback to File storage", stringBuilder);
+            }
+            else
+            {
+                _logger.LogError(
+                    $"S3 Media configuration validation failed with errors: {validationErrors} fallback to File storage", stringBuilder);
+            }
```
Any particular reason why logging here is environment-aware?

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/OrchardCore.Media.AmazonS3.csproj:
```diff
@@ -0,0 +1,21 @@
+    <Title>OrchardCore Media Amazon S3</Title>
+    <Description>$(OCCMSDescription)
+
+      Enables support for storing media files in Amazon S3 Bucket</Description>
```
```suggestion
      Enables support for storing media files in Amazon S3 Bucket.</Description>
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/OrchardCore.Media.AmazonS3.csproj:
```diff
@@ -0,0 +1,21 @@
+  </PropertyGroup>
+  
+  <ItemGroup>
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Admin.Abstractions\OrchardCore.Admin.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.FileStorage.AmazonS3\OrchardCore.FileStorage.AmazonS3.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Infrastructure.Abstractions\OrchardCore.Infrastructure.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Media.Core\OrchardCore.Media.Core.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Module.Targets\OrchardCore.Module.Targets.csproj" />
```
I don't think we need all of these.

Piedone reviewed (changes requested):
Awesome, thank you!

hishamco commented on src/docs/reference/modules/Media.Azure/README.md:
```diff
@@ -114,4 +114,4 @@ re-fetch the source file, as and when required, which the Media Cache Module wil
 !!! note
     The Media Feature is designed to support one storage provider at a time, whether that is
-    local File Storage, the default, or Azure Blob Storage.
+    local File Storage, the default, or Azure Blob Storage, or Amazon S3 Storage.
```
local File Storage (the default) ,  Azure Blob Storage or Amazon S3 Storage

neglectedvalue replied:
Changed, ty.

Piedone commented:
Please ask for a review (upper right corner) if you're done addressing all the feedback.

neglectedvalue commented:
I've looked at Azure Media module and I want to transfer some things into Amazon Media : templating for BucketName/ BasePath and ModularEvent so it would be possible to create S3 bucket automatically. Hope I'll finish it today.

Piedone commented on src/OrchardCore.Cms.Web/appsettings.json:
```diff
@@ -42,6 +42,20 @@
+    // See https://docs.orchardcore.net/en/latest/docs/reference/modules/Media.Amazon/#configuration to configure media storage in Amazon S3 Storage.
+    //"OrchardCore_Media_Amazon_S3": {
+    //  "BucketName": "",
+    // Credentials section needed only if Orchard will be hosted not in the AWS Cloud
+    // since in the AWS you don't need to store this info in the appsettings file, you 
+    // just need to set BucketName and BasePath.
```
```suggestion
    // Credentials section needed only if Orchard will be hosted not in the AWS Cloud
    // since in AWS you don't need to store this info in the appsettings file, you just
    // need to set BucketName and BasePath.
```

Piedone commented on src/docs/reference/modules/Media.Amazon/README.md:
```diff
@@ -0,0 +1,91 @@
+            "Credentials": {
+              "SecretKey": "",
+              "AccessKeyId": "",
+              "RegionEndpoint": ""
```
Please add some guidance to the docs on where to get this from. I tried the corresponding endpoints from https://docs.aws.amazon.com/general/latest/gr/s3.html, as well as using a more generic eu-central-1.amazonaws.com endpoint (for a Frankfurt-based bucket), but every single one of them cause "ArgumentException: Invalid region endpoint provided".

neglectedvalue replied:
I've added small comment about region in the doc file. Basically you need to use values from Region column, in your case just `eu-central-1`, but as I said I've added comment in doc.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsExtension.cs:
```diff
@@ -0,0 +1,90 @@
+
+    public static AwsStorageOptions BindConfiguration(this AwsStorageOptions options, IShellConfiguration shellConfiguration)
+    {
+        var section = shellConfiguration.GetSection("OrchardCore_Media_Amazon_S3");
```
In fact before we would have used `OrchardCore.Media.AmazonS3`.

But for compatibility under linux we needed to replace all `.` by `_`.

So maybe `OrchardCore_Media_AmazonS3`, only to be consistent with the module/feature name.

neglectedvalue replied:
Yeap, changed.

neglectedvalue commented on src/OrchardCore.Cms.Web/appsettings.json:
```diff
@@ -42,6 +42,20 @@
     //  "AllowedFileExtensions": [".jpg",".jpeg",".png",".gif",".ico",".svg",".pdf",".doc",".docx",".ppt",".pptx",".pps",".ppsx",".odt",".xls",".xlsx",".psd",".mp3",".m4a",".ogg",".wav",".mp4",".m4v",".mov",".wmv",".avi",".mpg",".ogv",".3gp"],
     //  "ContentSecurityPolicy": "default-src 'self'; style-src 'unsafe-inline'"
     //}
+    // See https://docs.orchardcore.net/en/latest/docs/reference/modules/Media.Amazon/#configuration to configure media storage in Amazon S3 Storage.
```
Done.

neglectedvalue commented on src/docs/reference/README.md:
```diff
@@ -66,6 +66,7 @@ Here's a categorized overview of all built-in Orchard Core features at a glance.
 - Media:
   - [Media](modules/Media/README.md)
   - [Media Slugify](modules/Media.Slugify/README.md)
+  - [Media Amazon](modules/Media.Amazon/README.md)
```
Added

Piedone commented on src/OrchardCore/OrchardCore.FileStorage.AmazonS3/AwsFileStorage.cs:
```diff
@@ -271,6 +271,7 @@ public async Task<string> CreateFileFromStreamAsync(string path, Stream inputStr
             {
                 BucketName = _options.BucketName,
                 Key = this.Combine(_basePrefix, path),
+                CannedACL = S3CannedACL.PublicRead,
```
Are you sure public read is necessary?

Piedone commented on src/docs/reference/modules/Media.AmazonS3/README.md:
```diff
@@ -0,0 +1,136 @@
+# Amazon S3 Media Storage (`OrchardCore.Media.AmazonS3`)
```
Please document what kind of bucket configuration is necessary (like ACLs).

Piedone commented on src/docs/reference/modules/Media.AmazonS3/README.md:
````diff
@@ -0,0 +1,136 @@
+}
+```
+
+Refer also to the [Configuration Section](../../core/Configuration/README.md),
+and the [Media Section](../Media/README.md) for other Media related configuration settings.
+
+There are two hosting options: inside and outside AWS Cloud.
+In case you are hosting Orchard Core inside AWS (EC2, EKS, etc.) you need to configure only `BucketName` and you may delete or comment out the other sections.
+
+In case you are hosting Orchard Core outside of AWS, you should fill the `Credentials` section or if you have AWS CLI installed and configured on your server you may specify only configured profile name (`default` if a profile name was not chosen during AWS CLI configuration).
+
+You can find region endpoints in the [Official AWS S3 Documentation](https://docs.aws.amazon.com/general/latest/gr/s3.html), see Region column. For example for Frankfurt region you should use `eu-central-1` 
+
+## AWS S3 Bucket Configuration
+
+By default, AWS 3S Bucket has limitations for newly uploaded files. If you want media files to be available from the outside of AWS, you should set up your bucket permissions.
+
+The simplest way of doing it is to add a policy:
````
```suggestion
Refer also to the [Configuration Section](../../core/Configuration/README.md),
and the [Media Section](../Media/README.md) for other Media related configuration settings.

There are two hosting options: inside and outside AWS Cloud.
In case you are hosting Orchard Core inside AWS (EC2, EKS, etc.) you need to configure only `BucketName` and you may delete or comment out the other sections.

In case you are hosting Orchard Core outside of AWS, you should fill the `Credentials` section or if you have AWS CLI installed and configured on your server you may specify only configured profile name (`default` if a profile name was not chosen during AWS CLI configuration).

You can find region endpoints in the [Official AWS S3 Documentation](https://docs.aws.amazon.com/general/latest/gr/s3.html), see Region column. For example for the Frankfurt region you should use `eu-central-1` 

## AWS S3 Bucket Configuration

By default, AWS 3S Bucket has limitations for newly uploaded files. If you want media files to be available from the outside of AWS, you should set up your bucket permissions.

The simplest way of doing it is to add a policy:
```

Piedone commented on src/docs/reference/modules/Media.AmazonS3/README.md:
````diff
@@ -0,0 +1,136 @@
+    ]
+}
+```
+After this policy will be added to your bucket permissions all newly added files will have Read permission and will be available from the outside of the Amazon Cloud.
+
+## Templating Configuration
+
+Optionally you may use liquid templating to further configure Amazon Media Storage, perhaps creating a bucket per tenant,
+or a single bucket with a base path per tenant.
+
+The `ShellSettings` property is made available to the liquid template.
+The `BucketName` property and the `BasePath` property are the only templatable properties.
+
+!!! note
+When templating the `BucketName`  using  `{{ ShellSettings.Name }}`, the tenant's name will be automatically lowercased, however, you must also make sure the `BucketName` conforms to other Amazon S3 naming conventions as set out in Amazon's documentation.
````
```suggestion
After this policy will be added to your bucket permissions all newly added files will have Read permission and will be available from the outside of the Amazon Cloud.

## Templating Configuration

Optionally you may use liquid templating to further configure Amazon Media Storage, perhaps creating a bucket per tenant,
or a single bucket with a base path per tenant.

The `ShellSettings` property is made available to the liquid template.
The `BucketName` property and the `BasePath` property are the only templatable properties.

!!! note
    When templating the `BucketName` using `{{ ShellSettings.Name }}`, the tenant's name will be automatically lowercased, however, you must also make sure the `BucketName` conforms to other Amazon S3 naming conventions as set out in Amazon's documentation.
```

Piedone commented on src/docs/reference/modules/Media.AmazonS3/README.md:
```diff
@@ -0,0 +1,136 @@
+!!! note
+When templating the `BucketName`  using  `{{ ShellSettings.Name }}`, the tenant's name will be automatically lowercased, however, you must also make sure the `BucketName` conforms to other Amazon S3 naming conventions as set out in Amazon's documentation.
+
+### Configuring a bucket per tenant.
```
```suggestion
### Configuring a bucket per tenant
```

Piedone commented on src/docs/reference/modules/Media.AmazonS3/README.md:
````diff
@@ -0,0 +1,136 @@
+}
+```
+
+### Configuring a single bucket, with a base folder per tenant.
````
```suggestion
### Configuring a single bucket, with a base folder per tenant
```

Piedone commented on src/docs/reference/modules/Media.AmazonS3/README.md:
```diff
@@ -0,0 +1,136 @@
+Optionally you may use liquid templating to further configure Amazon Media Storage, perhaps creating a bucket per tenant,
+or a single bucket with a base path per tenant.
+
+The `ShellSettings` property is made available to the liquid template.
```
Capitalize Liquid everywhere.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -0,0 +1,80 @@
+    // Local instance since it can be discarded once the startup is over
+    private readonly FluidParser _fluidParser = new FluidParser();
+
+    public AwsStorageOptionsConfiguration(
+        IShellConfiguration shellConfiguration,
+        ShellSettings shellSettings,
+        ILogger<AwsStorageOptions> logger
+    )
```
```suggestion
    public AwsStorageOptionsConfiguration(
        IShellConfiguration shellConfiguration,
        ShellSettings shellSettings,
        ILogger<AwsStorageOptions> logger)
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -0,0 +1,80 @@
+    private readonly ILogger _logger;
+
+    // Local instance since it can be discarded once the startup is over
+    private readonly FluidParser _fluidParser = new FluidParser();
```
```suggestion
    private readonly FluidParser _fluidParser = new();
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/CreateMediaS3BucketEvent.cs:
```diff
@@ -0,0 +1,73 @@
+    public CreateMediaS3BucketEvent(ShellSettings shellSettings,
+        IOptions<AwsStorageOptions> options,
+        IAmazonS3 amazonS3Client,
+        ILogger<CreateMediaS3BucketEvent> logger
+        )
```
```suggestion
        ILogger<CreateMediaS3BucketEvent> logger)
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/CreateMediaS3BucketEvent.cs:
```diff
@@ -0,0 +1,73 @@
+    {
+        if (_shellSettings.State != Environment.Shell.Models.TenantState.Uninitialized &&
+            !String.IsNullOrEmpty(_options.BucketName) &&
+            _options.CreateBucket)
```
This can be the first clause, because if it's `false` then nothing else to do.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/CreateMediaS3BucketEvent.cs:
```diff
@@ -0,0 +1,73 @@
+                    return;
+                }
+
+                var bucketRequest = new PutBucketRequest
```
I think the bucket should be created with Block Public Access.

Piedone reviewed (changes requested):
Thanks for adding Fluid support!

I know I'm pushing it, but probably you'd check out adding the S3 pair of `OrchardCore.DataProtection.Azure` next?

neglectedvalue commented:
I've added Public Access Block for newly created bucket and fixed typos in documentation.
Regarding DataProtection module - actually good idea but i've done small research and Amazon removed S3 DataProtection package and created [Amazon.AspNetCore.DataProtection.SSM](https://www.nuget.org/packages/Amazon.AspNetCore.DataProtection.SSM/) based on AWS System Manager. So there are two options: create a S3 DataProtection from a scratch or reuse  AWS SSM based DataProtection. I guess I'll do some POCs for both variants, but I guess it is better to create a new PR for that feature.

Piedone commented:
Yes, definitely a separate PR. Thank you!

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -53,6 +48,7 @@ private void ParseBucketName(AwsStorageOptions options, TemplateContext template
             var template = _fluidParser.Parse(options.BucketName);
 
+            options.BucketName = template.Render(templateContext, NullEncoder.Default);
             options.BucketName = options.BucketName.Replace("\r", String.Empty).Replace("\n", String.Empty);
```
```suggestion
            options.BucketName = template
                .Render(templateContext, NullEncoder.Default)
                .Replace("\r", String.Empty).Replace("\n", String.Empty);
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/CreateMediaS3BucketEvent.cs:
```diff
@@ -54,14 +52,30 @@ public override async Task ActivatingAsync()
                     return;
                 }
 
+                // Blocking public access for newly created bucket
```
```suggestion
                // Blocking public access for the newly created bucket.
```

Piedone reviewed (changes requested):
Apart from these and https://github.com/OrchardCMS/OrchardCore/pull/11738#discussion_r881072338 looks good.

Piedone commented:
Oh, and also address https://github.com/OrchardCMS/OrchardCore/pull/11738#discussion_r880861790, please.

Piedone commented on src/docs/reference/modules/Media.AmazonS3/README.md:
```diff
@@ -48,6 +49,13 @@ You can find region endpoints in the [Official AWS S3 Documentation](https://doc
 
 ## AWS S3 Bucket Configuration
 
+If `CreateBucket` was configured as `true` and `BucketName` is following official [Bucket naming rules](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html), than new bucket will be created.
+New bucket will be created without [Access Control Lists](https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html) due to security reasons, you may add needed ACLs manually:
+1) Open you bucket
+2) Go to the Permissions tab
+3) Edit `Block public access` 
```
```suggestion
If `CreateBucket` was configured as `true` and `BucketName` follows official [Bucket naming rules](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html), then a new bucket will be created.
The new bucket will be created without [Access Control Lists](https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html) due to security reasons. If you create the bucket manually then you need to do it with ACLs enabled. When using a previously created bucket, you may need to configure ACLs manually:

1. Open your bucket.
2. Go to the Permissions tab.
3. Edit "Block public access".
4. Tick "Block all public access".
```

Piedone reviewed (approved):
Anything you'd like to add, @hishamco @jtkech?

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -0,0 +1,77 @@
+
+            options.BucketName = template
+                .Render(templateContext, NullEncoder.Default)
+                .Replace("\r", String.Empty).Replace("\n", String.Empty);
```
Why not just trim. If there are space in the middle of the string they should be intentional. I could understand that having a new line in a fluid template might be hardly noticeable though.

hishamco replied:
Can we use `Environment.NewLine` instead?

neglectedvalue replied:
Instead of "\n"? So result will be smth like :
`            options.BucketName = template
                .Render(templateContext, NullEncoder.Default)
                .Replace("\r", String.Empty)
                .Replace(System.Environment.NewLine, String.Empty)
                .Trim();`

hishamco replied:
`Environment.NewLine` insstead of `\r\n`

hishamco replied:
```csharp
options.BucketName = template .Render(templateContext, NullEncoder.Default) .Replace(Environment.NewLine, String.Empty)
```

hishamco replied:
```csharp
options.BucketName = template .Render(templateContext, NullEncoder.Default) .Replace(Environment.NewLine, String.Empty)
```

neglectedvalue replied:
Added, I've also added Trim() as Sebastien mention

jtkech replied:
@hishamco

Yes `Environment.NewLine` is useful to generate strings on a given platform.

But not sure it is a good idea to cleanup strings that may come from different sources.

I would have let the replace of `\r` and `\n`.

neglectedvalue replied:
So guys, should i rollback my latest commit? :)

jtkech replied:
I think it is safer, but keep the trim

neglectedvalue replied:
Done.

hishamco replied:
> I would have let the replace of \r and \n.

Is there anything wrong with `Environment.NewLine`?

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -0,0 +1,77 @@
+            var template = _fluidParser.Parse(options.BasePath);
+
+            options.BasePath = template.Render(templateContext, NullEncoder.Default);
+            options.BasePath = options.BasePath.Replace("\r", String.Empty).Replace("\n", String.Empty);
```
Same

hishamco commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -50,7 +50,8 @@ private void ParseBucketName(AwsStorageOptions options, TemplateContext template
             options.BucketName = template
                 .Render(templateContext, NullEncoder.Default)
-                .Replace("\r", String.Empty).Replace("\n", String.Empty);
+                .Replace(System.Environment.NewLine, String.Empty)
```
Last thing add `using System` instead

neglectedvalue replied:
There is a conflict in Environment naming between OrchardCore.Environment and System, so it's either : 
System.Environment.NewLine
or 
using static System.Environment;
and
```C#
 options.BucketName = template
                .Render(templateContext, NullEncoder.Default)
                .Replace("\r", String.Empty).Replace("\n", String.Empty);
                .Replace(NewLine, String.Empty)
```

hishamco commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -65,8 +66,10 @@ private void ParseBasePath(AwsStorageOptions options, TemplateContext templateCo
-            options.BasePath = options.BasePath.Replace("\r", String.Empty).Replace("\n", String.Empty);
+            options.BasePath = template
+                .Render(templateContext, NullEncoder.Default)
+                .Replace(System.Environment.NewLine, String.Empty)
```
Same here

hishamco commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -50,7 +50,8 @@ private void ParseBucketName(AwsStorageOptions options, TemplateContext template
                 .Render(templateContext, NullEncoder.Default)
-                .Replace("\r", String.Empty).Replace("\n", String.Empty);
+                .Replace(System.Environment.NewLine, String.Empty)
+                .Trim();
```
I don't think we need to trim the string if we remove the new line, unless there' another white-space. Just leave it as its

hishamco commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AwsStorageOptionsConfiguration.cs:
```diff
@@ -51,8 +52,7 @@ private void ParseBucketName(AwsStorageOptions options, TemplateContext template
             options.BucketName = template
                 .Render(templateContext, NullEncoder.Default)
                 .Replace("\r", String.Empty)
```
Why we still replace `\r`?

neglectedvalue replied:
just in case some one saved config in CRLF mode and deployed it in unix based OS

jtkech replied:
> just in case some one saved config in CRLF mode and deployed it in unix based OS

@neglectedvalue you're right, thanks for your patience.

Then @hishamco what happens if

> just in case some one saved config in LF mode and deployed it in windows based OS.

Your suggestion does the following on windows, which doesn't replace `\n` unix newlines.

    options.BucketName = template
        .Render(templateContext, NullEncoder.Default)
        .Replace("\r", String.Empty)
        .Replace(\r\n, String.Empty);

> We already used quite often https://github.com/OrchardCMS/OrchardCore/search?q=NewLine

As said, useful to generate strings but may be wrong to process strings from different sources.

hishamco commented:
Just a note to @agriffard please add it as a contributor once the PR is merged

Piedone commented:
@jtkech anything you'd like to add or would you approve, please?

jtkech commented:
Related to the comments I did, that's okay for me.
Didn't look at all things, but LGTM so I will approve.

Piedone commented:
Congrats @neglectedvalue and thank you for this contribution!
